[1.0-dev] Store indices in CartesianRange

timholy opened the pull request:
The main advantage of this change is that you no longer lose type information from `CartesianRange(indices(A))`. The types of the specific AbstractUnitRanges are important for `similar`, if you're using non-1 indices. It's also kind of nice to be able to write `CartesianRange{3}` rather than `CartesianRange{CartesianIndex{3}}`.

Because this changes the parametrization of an exported type, and also introduces a new depwarn, this is queued for after 0.6.

CC @JKrehl.

JKrehl commented:
Not having checked all the ends this change seeps into, one question:
Why are the ranges forced to use `Int`? There are credible use-cases for using other data types for indexing an array (sparse arrays of low occupancy, lots of small arrays).

timholy commented:
Because the indices in `CartesianIndex` are `Int`. I'd rather get the conversion error in the outer loop than the inner.

JKrehl commented:
And also changing `CartesianIndex` to be based on `NTuple{<:Integer}` is maybe to much for this PR.

JKrehl commented:
@timholy 
Looks good to me except the little utility/magic function `ur_int`, wouldn't it be better to use `convert(AbstractUnitRange{Int}, ...)` for this?

timholy commented:
That's not defined currently, but heck yeah, that's a better way to "spell" that conversion. Updated.

tkelman commented:
bump, revisit?

timholy commented:
Yup, on my agenda. I'm a bit focused on finishing #22210 now, but once the dust settles there I'll rebase this.

tkelman commented on base/deprecated.jl:
```diff
@@ -1546,4 +1546,10 @@ end
 # END 0.7 deprecations
 
 # BEGIN 1.0 deprecations
+function CartesianRange{N}(start::CartesianIndex{N}, stop::CartesianIndex{N})
```
0.7 section, right?

timholy committed:
Store indices in CartesianRange

The main advantage of this change is that you no longer lose type information from `CartesianRange(indices(A))`. The types of the specific AbstractUnitRanges are important for `similar`, etc.
